pipeline.py

```diff
@@ -0,0 +1,231 @@
+    return processed_bold_np
+
+def get_bold_in_roi(bold, roi_mask=nib.load('subj1/mask4_vt.nii.gz')):
+
+    return bold.get_fdata()[roi_mask.get_fdata() == 1, :]
+
+# Plot anatomy
+#plotting.plot_anat(anat)
+
+
+def searchlight(center_voxels, radius, search_grid):
+    """This is a searchlight function that returns an array for each center voxel corresponding to
+    all search_grid locations that are within radius (euclidean distance) from the center voxel
+
+    Parameters
+    ----------
+    center_voxels : list
+        a list or array of voxel locations around which seearch is performed
+    radius: float
+        search radius: locations within this radius from each center voxel are included in return
+    search_grid: list
+        all possible locations in 3D coordinates, in list format
+
+    Returns
+    -------
+    dict
+        dictionary of lists, where each key is a center voxel and items are lists of locations inside
+        the search radius from that center voxel in 2D array format: line is a voxel, columns are coordinates.
+
+    """
+    search_results = {}
+    for center_voxel in center_voxels:
+        distances = scipy.spatial.distance.cdist(search_grid, np.array([center_voxel]))
+        found_locations = np.array(search_grid)[distances.flatten() <= radius]
+        search_results[center_voxel] = found_locations
+
+
+    return search_results
+
+
+def build_model_RDM(labels):
+
+    model_RDM = np.zeros((len(labels.index), len(labels.index)))
+    labels_sorted = labels.sort_values('labels')
+
+
+
+def remove_rest(data, labels):
+    """Removes 'rest' labeled data.
+
+    Parameters
+    ----------
+    data : array-like
+        data to be cleaned
+
+    labels : Pandas dataframe
+        labels corresponding to last axis in data
+
+    Returns
+    -------
+    array-like
+        data that has been cleaned of data corresponding to 'rest' labels
+
+    Pandas dataframe
+        labels that correspond to returned data
+
+    """
+
+    clean_data = data[:, :, :, labels.query('labels != "rest" ').index]
+    clean_labels = labels[labels.labels != 'rest']
+
+    return clean_data, clean_labels
+
+def create_model_RDM(labels):
+    model_RDM = np.zeros((len(labels.index), len(labels.index)))
+
+    for i in range(len(labels_sorted.labels)):
+        for j in range(len(labels_sorted.labels)):
+            model_RDM[i, j] = (labels_sorted.labels.iloc[i] == labels_sorted.labels.iloc[j])
+
+    return model_RDM
+
+
+def flatten_voxels(data, locations, image):
+    flat = np.array([])
+    for loc in locations:
+        flat = np.append(flat, data[loc[0], loc[1], loc[2], image])
+
+    return flat
+
+
+def create_bold_RDM(data, location, ROI_searchlights):
+    n_images = len(data[0, 0, 0, :])
+    RDM = np.zeros((n_images, n_images))
+    search_locs = ROI_searchlights[location]
+
+    progress_bar = Bar('Single RDM progress', max=n_images)
+    for i in range(n_images):
+        for j in range(n_images):
```

Comment:
this double for-loop is super slow. Take a look at scipy's [`pdist`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.pdist.html) function, in combination with [`squareform`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.squareform.html):

if data is a 2d array (n_images x n_voxels_in_searchlight_patch):
```python
from scipy.spatial.distance import pdist, squareform
RDM = squareform(pdist(data, method='correlation'))
```